File: impl/go/scanner.go
package twill

import (
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// Source scanning: file discovery, candidate extraction, and incremental
// scans (SPEC §13).

var ignoredDirectories = map[string]bool{
	".git": true, ".hg": true, ".jj": true, ".next": true, ".parcel-cache": true,
	".pnpm-store": true, ".svelte-kit": true, ".svn": true, ".turbo": true,
	".venv": true, ".vercel": true, ".yarn": true, "__pycache__": true,
	"node_modules": true, "venv": true,
}

var ignoredExtensions = map[string]bool{
	"less": true, "lock": true, "sass": true, "scss": true, "styl": true, "log": true,
	// Images
	"png": true, "jpg": true, "jpeg": true, "gif": true, "webp": true, "avif": true,
	"ico": true, "bmp": true, "tif": true, "tiff": true, "heic": true, "psd": true,
	// Audio
	"mp3": true, "wav": true, "ogg": true, "flac": true, "aac": true, "m4a": true,
	// Video
	"mp4": true, "webm": true, "mov": true, "avi": true, "mkv": true, "m4v": true,
	// Archives
	"zip": true, "gz": true, "tar": true, "tgz": true, "bz2": true, "xz": true,
	"7z": true, "rar": true,
	// Fonts
	"woff": true, "woff2": true, "ttf": true, "otf": true, "eot": true,
	// Executables and binaries
	"exe": true, "dll": true, "so": true, "dylib": true, "bin": true, "wasm": true,
	"pdf": true, "class": true, "jar": true, "pyc": true, "o": true, "a": true,
}

var ignoredFiles = map[string]bool{
	"package-lock.json": true, "pnpm-lock.yaml": true, "bun.lockb": true,
	".gitignore": true, ".env": true,
}

func isIgnoredFileName(name string) bool {
	if ignoredFiles[name] || strings.HasPrefix(name, ".env.") {
		return true
	}
	if dot := strings.LastIndex(name, "."); dot > 0 {
		if ignoredExtensions[strings.ToLower(name[dot+1:])] {
			return true
		}
	}
	return false
}

// IsGlob reports whether pattern contains glob metacharacters.
func IsGlob(pattern string) bool { return strings.ContainsAny(pattern, "*?[{") }

// GlobToRegexp converts an absolute posix glob (`**`, `*`, `?`, `{a,b}`)
// into a regular expression matching whole paths.
func GlobToRegexp(glob string) *regexp.Regexp {
	var b strings.Builder
	b.WriteString("^")
	braces := 0
	for i := 0; i < len(glob); i++ {
		c := glob[i]
		switch {
		case c == '*':
			if i+1 < len(glob) && glob[i+1] == '*' {
				i++
				if i+1 < len(glob) && glob[i+1] == '/' {
					i++
					b.WriteString("(?:.*/)?")
				} else {
					b.WriteString(".*")
				}
			} else {
				b.WriteString("[^/]*")
			}
		case c == '?':
			b.WriteString("[^/]")
		case c == '{':
			braces++
			b.WriteString("(?:")
		case c == '}' && braces > 0:
			braces--
			b.WriteString(")")
		case c == ',' && braces > 0:
			b.WriteString("|")
		case c == '\\' && i+1 < len(glob):
			i++
			b.WriteString(regexp.QuoteMeta(string(glob[i])))
		default:
			b.WriteString(regexp.QuoteMeta(string(c)))
		}
	}
	b.WriteString("$")
	return regexp.MustCompile(b.String())
}

// exclusion is a compiled exclusion from a negated source.
type exclusion struct {
	// Every path at or under this directory is excluded.
	directory string
	// Or: files matching this pattern (absolute, posix).
	pattern *regexp.Regexp
}

type scopedIgnore struct {
	directory string
	matcher   *Gitignore
}

func isDirectoryPath(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}

func isFilePath(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.Mode().IsRegular()
}

func absPath(path string) string {
	abs, err := filepath.Abs(path)
	if err != nil {
		return filepath.Clean(path)
	}
	return abs
}

// Scanner enumerates source files from globs and auto-detection rules,
// extracts candidates, and tracks modification times for incremental scans.
type Scanner struct {
	sources    []SourceEntry
	candidates *candidateSet
	mtimes     map[string]int64
	exclusions []exclusion
	compiled   bool

	// ScannedFiles are the files read by the most recent Scan.
	ScannedFiles []string
}

// NewScanner creates a scanner for the given sources.
func NewScanner(sources []SourceEntry) *Scanner {
	return &Scanner{sources: sources, candidates: newCandidateSet(), mtimes: map[string]int64{}}
}

// Sources returns the sources this scanner was created with.
func (s *Scanner) Sources() []SourceEntry { return s.sources }

// Bases returns the distinct absolute base directories of the positive sources.
func (s *Scanner) Bases() []string {
	seen := map[string]bool{}
	var bases []string
	for _, source := range s.sources {
		if source.Negated {
			continue
		}
		base := absPath(source.Base)
		if !seen[base] {
			seen[base] = true
			bases = append(bases, base)
		}
	}
	return bases
}

func (s *Scanner) exclusionsFor() []exclusion {
	if s.compiled {
		return s.exclusions
	}
	for _, source := range s.sources {
		if !source.Negated {
			continue
		}
		target := absPath(filepath.Join(source.Base, source.Pattern))
		if !IsGlob(source.Pattern) {
			if isDirectoryPath(target) {
				s.exclusions = append(s.exclusions, exclusion{directory: target})
				continue
			}
			s.exclusions = append(s.exclusions, exclusion{pattern: regexp.MustCompile("^" + regexp.QuoteMeta(filepath.ToSlash(target)) + "$")})
			continue
		}
		s.exclusions = append(s.exclusions, exclusion{pattern: GlobToRegexp(filepath.ToSlash(target))})
	}
	s.compiled = true
	return s.exclusions
}

func (s *Scanner) isExcluded(path string, exclusions []exclusion) bool {
	posix := filepath.ToSlash(path)
	for _, e := range exclusions {
		if e.directory != "" {
			if path == e.directory || strings.HasPrefix(path, e.directory+string(filepath.Separator)) {
				return true
			}
		} else if e.pattern != nil && e.pattern.MatchString(posix) {
			return true
		}
	}
	return false
}

type fileSet struct {
	seen map[string]bool
	list []string
}

func (f *fileSet) add(path string) {
	if f.seen == nil {
		f.seen = map[string]bool{}
	}
	if !f.seen[path] {
		f.seen[path] = true
		f.list = append(f.list, path)
	}
}

// Files enumerates every file covered by the sources.
func (s *Scanner) Files() []string {
	exclusions := s.exclusionsFor()
	found := &fileSet{}
	for _, source := range s.sources {
		if source.Negated {
			continue
		}
		base := absPath(source.Base)
		if source.Pattern == "**/*" {
			s.walk(base, nil, exclusions, found)
			continue
		}
		target := absPath(filepath.Join(base, source.Pattern))
		if !IsGlob(source.Pattern) {
			if isDirectoryPath(target) {
				s.walk(target, nil, exclusions, found)
			} else if isFilePath(target) && !s.isExcluded(target, exclusions) {
				found.add(target)
			}
			continue
		}
		s.glob(target, exclusions, found)
	}
	return found.list
}

func (s *Scanner) walk(directory string, ignores []scopedIgnore, exclusions []exclusion, found *fileSet) {
	scoped := ignores
	if content, err := os.ReadFile(filepath.Join(directory, ".gitignore")); err == nil {
		matcher := ParseGitignore(string(content))
		if matcher.Size() > 0 {
			scoped = append(append([]scopedIgnore{}, ignores...), scopedIgnore{directory, matcher})
		}
	}

	entries, err := os.ReadDir(directory)
	if err != nil {
		return
	}
	for _, entry := range entries {
		path := filepath.Join(directory, entry.Name())
		isDir := entry.IsDir()
		if entry.Type()&os.ModeSymlink != 0 {
			isDir = isDirectoryPath(path)
		}
		if isDir {
			if ignoredDirectories[entry.Name()] || s.isGitignored(path, true, scoped) || s.isExcluded(path, exclusions) {
				continue
			}
			s.walk(path, scoped, exclusions, found)
			continue
		}
		if isIgnoredFileName(entry.Name()) || s.isGitignored(path, false, scoped) || s.isExcluded(path, exclusions) {
			continue
		}
		found.add(path)
	}
}

func (s *Scanner) isGitignored(path string, isDir bool, ignores []scopedIgnore) bool {
	for _, ignore := range ignores {
		rel, err := filepath.Rel(ignore.directory, path)
		if err != nil {
			continue
		}
		if ignore.matcher.Ignores(filepath.ToSlash(rel), isDir) {
			return true
		}
	}
	return false
}

func (s *Scanner) glob(target string, exclusions []exclusion, found *fileSet) {
	absolute := filepath.ToSlash(target)
	pattern := GlobToRegexp(absolute)
	// Walk from the first non-glob segment of the absolute pattern.
	var fixed []string
	for _, segment := range strings.Split(absolute, "/") {
		if IsGlob(segment) {
			break
		}
		fixed = append(fixed, segment)
	}
	start := strings.Join(fixed, "/")
	if start == "" {
		start = "/"
	}
	start = filepath.FromSlash(start)
	if isFilePath(start) {
		if !s.isExcluded(start, exclusions) {
			found.add(start)
		}
		return
	}
	s.walkAll(start, pattern, exclusions, found)
}

func (s *Scanner) walkAll(directory string, pattern *regexp.Regexp, exclusions []exclusion, found *fileSet) {
	entries, err := os.ReadDir(directory)
	if err != nil {
		return
	}
	for _, entry := range entries {
		path := filepath.Join(directory, entry.Name())
		isDir := entry.IsDir()
		if entry.Type()&os.ModeSymlink != 0 {
			isDir = isDirectoryPath(path)
		}
		if isDir {
			if entry.Name() == ".git" || s.isExcluded(path, exclusions) {
				continue
			}
			s.walkAll(path, pattern, exclusions, found)
			continue
		}
		if !pattern.MatchString(filepath.ToSlash(path)) || s.isExcluded(path, exclusions) {
			continue
		}
		found.add(path)
	}
}

func (s *Scanner) read(path string) bool {
	content, err := os.ReadFile(path)
	if err != nil {
		return false
	}
	extractInto(string(content), s.candidates)
	return true
}

// Scan walks every source, reads files whose modification time changed
// since the last scan (all files on the first scan), and returns the full
// deduplicated candidate set seen so far.
func (s *Scanner) Scan() []string {
	read := []string{}
	for _, path := range s.Files() {
		info, err := os.Stat(path)
		if err != nil {
			continue
		}
		mtime := info.ModTime().UnixNano()
		if previous, ok := s.mtimes[path]; ok && previous == mtime {
			continue
		}
		s.mtimes[path] = mtime
		if s.read(path) {
			read = append(read, path)
		}
	}
	s.ScannedFiles = read
	return s.candidates.slice()
}

// ScanFiles reads only the given files and returns the candidates not seen
// before.
func (s *Scanner) ScanFiles(changed []string) []string {
	before := s.candidates.len()
	for _, path := range changed {
		info, err := os.Stat(path)
		if err != nil || !info.Mode().IsRegular() {
			continue
		}
		s.mtimes[path] = info.ModTime().UnixNano()
		s.read(path)
	}
	return append([]string{}, s.candidates.list[before:]...)
}

// Candidates returns every candidate seen so far.
func (s *Scanner) Candidates() []string { return s.candidates.slice() }

File: impl/ocaml/src/scanner.ml
(* Source scanning: file discovery, candidate extraction, and incremental
   scans (SPEC §13). *)

open Utils

let ignored_directories =
  [ ".git"; ".hg"; ".jj"; ".next"; ".parcel-cache"; ".pnpm-store"; ".svelte-kit"; ".svn"; ".turbo"; ".venv"; ".vercel"; ".yarn";
    "__pycache__"; "node_modules"; "venv" ]

let ignored_extensions =
  fields
    "less lock sass scss styl log png jpg jpeg gif webp avif ico bmp tif tiff heic psd mp3 wav ogg flac aac m4a mp4 webm mov avi \
     mkv m4v zip gz tar tgz bz2 xz 7z rar woff woff2 ttf otf eot exe dll so dylib bin wasm pdf class jar pyc o a"

let ignored_files = [ "package-lock.json"; "pnpm-lock.yaml"; "bun.lockb"; ".gitignore"; ".env" ]

let is_ignored_file_name name =
  List.mem name ignored_files || has_prefix name ".env."
  ||
  match String.rindex_opt name '.' with
  | Some dot when dot > 0 -> List.mem (String.lowercase_ascii (after name (dot + 1))) ignored_extensions
  | _ -> false

(* Whether a pattern contains glob metacharacters. *)
let is_glob pattern = List.exists (fun c -> String.contains pattern c) [ '*'; '?'; '['; '{' ]

(* Converts an absolute posix glob (`**`, `*`, `?`, `{a,b}`) into a regular
   expression matching whole paths. *)
let glob_to_regexp glob =
  let b = Buffer.create 64 in
  Buffer.add_string b "^";
  let n = String.length glob in
  let braces = ref 0 in
  let i = ref 0 in
  while !i < n do
    (match glob.[!i] with
    | '*' ->
        if !i + 1 < n && glob.[!i + 1] = '*' then begin
          incr i;
          if !i + 1 < n && glob.[!i + 1] = '/' then (incr i; Buffer.add_string b "\\(.*/\\)?") else Buffer.add_string b ".*"
        end
        else Buffer.add_string b "[^/]*"
    | '?' -> Buffer.add_string b "[^/]"
    | '{' ->
        incr braces;
        Buffer.add_string b "\\("
    | '}' when !braces > 0 ->
        decr braces;
        Buffer.add_string b "\\)"
    | ',' when !braces > 0 -> Buffer.add_string b "\\|"
    | '\\' when !i + 1 < n ->
        incr i;
        Buffer.add_string b (Str.quote (String.make 1 glob.[!i]))
    | c -> Buffer.add_string b (Str.quote (String.make 1 c)));
    incr i
  done;
  Buffer.add_string b "$";
  Str.regexp (Buffer.contents b)

let matches re s = Str.string_match re s 0

(* Normalizes `.` and `..` segments and repeated separators. *)
let clean path =
  let absolute = has_prefix path "/" in
  let rec go acc = function
    | [] -> List.rev acc
    | ("" | ".") :: rest -> go acc rest
    | ".." :: rest -> (
        match acc with
        | x :: acc' when x <> ".." -> go acc' rest
        | _ -> if absolute then go acc rest else go (".." :: acc) rest)
    | p :: rest -> go (p :: acc) rest
  in
  let joined = String.concat "/" (go [] (String.split_on_char '/' path)) in
  if absolute then "/" ^ joined else if joined = "" then "." else joined

let abs_path path = clean (if Filename.is_relative path then Filename.concat (Sys.getcwd ()) path else path)
let join a b = if Filename.is_relative b then clean (Filename.concat a b) else clean b
let is_directory_path path = try Sys.is_directory path with Sys_error _ -> false
let is_file_path path = try (Unix.stat path).st_kind = Unix.S_REG with Unix.Unix_error _ -> false
let mtime path = try Some (Unix.stat path).st_mtime with Unix.Unix_error _ -> None
let read_file path = try Some (In_channel.with_open_bin path In_channel.input_all) with Sys_error _ -> None

let entries directory =
  match Sys.readdir directory with
  | names ->
      let names = Array.to_list names in
      List.sort compare names
  | exception Sys_error _ -> []

(* A compiled exclusion from a negated source: every path at or under a
   directory, or files matching a pattern. *)
type exclusion = Excluded_directory of string | Excluded_pattern of Str.regexp

type scoped_ignore = { directory : string; matcher : Gitignore.t }

type t = {
  sources : Directives.source_entry list;
  candidates : Extract.candidate_set;
  mtimes : (string, float) Hashtbl.t;
  mutable exclusions : exclusion list option;
  (* The files read by the most recent [scan]. *)
  mutable scanned_files : string list;
}

let create sources =
  { sources; candidates = Extract.new_candidate_set (); mtimes = Hashtbl.create 256; exclusions = None; scanned_files = [] }

let sources s = s.sources

(* The distinct absolute base directories of the positive sources. *)
let bases s =
  List.fold_left
    (fun acc (source : Directives.source_entry) ->
      if source.negated then acc
      else
        let base = abs_path source.base in
        if List.mem base acc then acc else acc @ [ base ])
    [] s.sources

let exclusions_for s =
  match s.exclusions with
  | Some e -> e
  | None ->
      let e =
        List.filter_map
          (fun (source : Directives.source_entry) ->
            if not source.negated then None
            else
              let target = abs_path (join source.base source.pattern) in
              if not (is_glob source.pattern) then
                if is_directory_path target then Some (Excluded_directory target)
                else Some (Excluded_pattern (Str.regexp ("^" ^ Str.quote target ^ "$")))
              else Some (Excluded_pattern (glob_to_regexp target)))
          s.sources
      in
      s.exclusions <- Some e;
      e

let is_excluded path exclusions =
  List.exists
    (function
      | Excluded_directory d -> path = d || has_prefix path (d ^ "/")
      | Excluded_pattern re -> matches re path)
    exclusions

(* An insertion-ordered file set. *)
type file_set = { fseen : (string, unit) Hashtbl.t; mutable flist_rev : string list }

let add_file f path =
  if not (Hashtbl.mem f.fseen path) then begin
    Hashtbl.replace f.fseen path ();
    f.flist_rev <- path :: f.flist_rev
  end

let is_dir_entry path =
  match Unix.lstat path with
  | { Unix.st_kind = Unix.S_DIR; _ } -> true
  | { Unix.st_kind = Unix.S_LNK; _ } -> is_directory_path path
  | _ -> false
  | exception Unix.Unix_error _ -> false

let is_gitignored path is_dir ignores =
  List.exists
    (fun ignore ->
      let prefix = ignore.directory ^ "/" in
      let rel = if has_prefix path prefix then after path (String.length prefix) else path in
      Gitignore.ignores ignore.matcher rel is_dir)
    ignores

let rec walk directory ignores exclusions found =
  let scoped =
    match read_file (Filename.concat directory ".gitignore") with
    | Some content ->
        let matcher = Gitignore.parse content in
        if Gitignore.size matcher > 0 then ignores @ [ { directory; matcher } ] else ignores
    | None -> ignores
  in
  List.iter
    (fun name ->
      let path = Filename.concat directory name in
      if is_dir_entry path then begin
        if List.mem name ignored_directories || is_gitignored path true scoped || is_excluded path exclusions then ()
        else walk path scoped exclusions found
      end
      else if is_ignored_file_name name || is_gitignored path false scoped || is_excluded path exclusions then ()
      else add_file found path)
    (entries directory)

let rec walk_all directory pattern exclusions found =
  List.iter
    (fun name ->
      let path = Filename.concat directory name in
      if is_dir_entry path then begin
        if name = ".git" || is_excluded path exclusions then () else walk_all path pattern exclusions found
      end
      else if (not (matches pattern path)) || is_excluded path exclusions then ()
      else add_file found path)
    (entries directory)

let glob_files target exclusions found =
  let pattern = glob_to_regexp target in
  (* Walk from the first non-glob segment of the absolute pattern. *)
  let rec fixed acc = function [] -> List.rev acc | seg :: rest -> if is_glob seg then List.rev acc else fixed (seg :: acc) rest in
  let start = String.concat "/" (fixed [] (String.split_on_char '/' target)) in
  let start = if start = "" then "/" else start in
  if is_file_path start then (if not (is_excluded start exclusions) then add_file found start)
  else walk_all start pattern exclusions found

(* Enumerates every file covered by the sources. *)
let files s =
  let exclusions = exclusions_for s in
  let found = { fseen = Hashtbl.create 256; flist_rev = [] } in
  List.iter
    (fun (source : Directives.source_entry) ->
      if source.negated then ()
      else
        let base = abs_path source.base in
        if source.pattern = "**/*" then walk base [] exclusions found
        else
          let target = abs_path (join base source.pattern) in
          if not (is_glob source.pattern) then begin
            if is_directory_path target then walk target [] exclusions found
            else if is_file_path target && not (is_excluded target exclusions) then add_file found target
          end
          else glob_files target exclusions found)
    s.sources;
  List.rev found.flist_rev

let read s path =
  match read_file path with
  | Some content ->
      Extract.extract_into content s.candidates;
      true
  | None -> false

(* Walks every source, reads files whose modification time changed since
   the last scan (all files on the first scan), and returns the full
   deduplicated candidate set seen so far. *)
let scan s =
  let read_files =
    List.filter
      (fun path ->
        match mtime path with
        | None -> false
        | Some m -> (
            match Hashtbl.find_opt s.mtimes path with
            | Some previous when previous = m -> false
            | _ ->
                Hashtbl.replace s.mtimes path m;
                read s path))
      (files s)
  in
  s.scanned_files <- read_files;
  Extract.to_list s.candidates

(* Reads only the given files and returns the candidates not seen before. *)
let scan_files s changed =
  let before = s.candidates.size in
  List.iter
    (fun path ->
      if is_file_path path then begin
        (match mtime path with Some m -> Hashtbl.replace s.mtimes path m | None -> ());
        ignore (read s path)
      end)
    changed;
  List.filteri (fun i _ -> i >= before) (Extract.to_list s.candidates)

(* Every candidate seen so far. *)
let candidates s = Extract.to_list s.candidates
